backend/src/v2/driver/driver.go

```diff
@@ -1118,6 +1082,136 @@ var accessModeMap = map[string]k8score.PersistentVolumeAccessMode{
 	"ReadWriteOncePod": k8score.ReadWriteOncePod,
 }
 
+// createPVCandPublish() creates the PVC and publishes the execution to MLMD, regardless of the
+// creation succeeds or not. However, if we failed to initiate the k8s client, we cannot
+// publish the execution.
+func createPVCandPublish(
+	inputs *pipelinespec.ExecutorInput_Inputs,
+	namespace string,
+	mlmd *metadata.Client,
+	ctx context.Context,
+	execution *metadata.Execution,
+) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("failed to create pvc and publish execution %s: %w", execution.TaskName(), err)
+		}
+	}()
+	// If we cannot create Kubernetes client, we cannot publish this execution
+	k8sClient, err := createK8sClient()
+	if err != nil {
+		return fmt.Errorf("cannot generate k8s clientset: %w", err)
+	}
+
+	var outputParameters map[string]*structpb.Value
+	status := pb.Execution_FAILED
+	defer func() {
+		// We publish the execution regardless of whether PVC creation succeeds or not
+		perr := publishDriverExecution(k8sClient, mlmd, ctx, execution, outputParameters, nil, status)
+		if perr != nil && err != nil {
+			err = fmt.Errorf("failed to publish driver execution: %w. Also failed to create PVC: %w", perr, err)
+		} else if perr != nil {
+			err = fmt.Errorf("failed to publish driver execution: %w", perr)
+		}
+	}()
+
+	pvcName, err := createPVC(k8sClient, inputs, namespace)
+	if err != nil {
+		return err
+	}
+	outputParameters = map[string]*structpb.Value{
+		"name": structpb.NewStringValue(pvcName),
+	}
+	status = pb.Execution_COMPLETE
+	return nil
+}
+
+// deletePVCandPublish() deletes the PVC and publishes the execution to MLMD, regardless of the
+// creation succeeds or not. However, if we failed to initiate the k8s client, we cannot
+// publish the execution.
+func deletePVCandPublish(
```

Comment:
The two methods create and delete are mostly duplicate, can them be reduced into one?

Reply:
I think you have a good point here. If we add more driver only functions in the future, it would be cumbersome to have one function for each. I will change the code accordingly.